File: cafe-order-checker.cs
public static bool IsFirstComeFirstServed(int[] takeOutOrders, int[] dineInOrders, int[] servedOrders)
{
    int takeOutOrdersIndex = 0;
    int dineInOrdersIndex = 0;

    foreach (var order in servedOrders)
    {
        if (takeOutOrdersIndex < takeOutOrders.Length && order == takeOutOrders[takeOutOrdersIndex])
        {
            takeOutOrdersIndex++;
        }
        else if (dineInOrdersIndex < dineInOrders.Length && order == dineInOrders[dineInOrdersIndex])
        {
            dineInOrdersIndex++;
        }
        else
        {
            return false;
        }
    }

    if (dineInOrdersIndex != dineInOrders.Length || takeOutOrdersIndex != takeOutOrders.Length) {
        return false;
    }

    return true;
}

File: cake/cafe-order-checker.cpp
  bool isFirstComeFirstServed(const vector<int>& takeOutOrders,
                            const vector<int>& dineInOrders,
                            const vector<int>& servedOrders)
{
    auto takeOutOrdersIter = takeOutOrders.cbegin();
    auto dineInOrdersIter = dineInOrders.cbegin();

    for (int order : servedOrders) {
        if (takeOutOrdersIter != takeOutOrders.cend() && order == *takeOutOrdersIter) {
            ++takeOutOrdersIter;
        }

        else if (dineInOrdersIter != dineInOrders.cend() && order == *dineInOrdersIter) {
            ++dineInOrdersIter;
        }

        else {
            return false;
        }
    }

    if (dineInOrdersIter != dineInOrders.cend() || takeOutOrdersIter != takeOutOrders.cend()) {
        return false;
    }

    return true;
}